### PR TITLE
Remove lodash functions

### PR DESCRIPTION
**Pull request by a-tarasyuk**
cc @deregtd @berickson1

### CODE REVIEW

**Comment by berickson1** [comment]
LGTM. Thanks! @deregtd @ms-markda - can you take a quick peek, I don't see anything wrong here but would like a couple more eyes to make sure there isn't any unexpected behavioral changes

File: src/ComponentBase.ts [hunk 1]
```diff
@@ -153,20 +152,20 @@ export abstract class ComponentBase<P extends {}, S extends _.Dictionary<any>> e
 
     // Subclasses may override, but _MUST_ call super.
     componentWillUnmount(): void {
-        _.forEach(this._handledSubscriptions, (subscription: StoreSubscriptionInternal<P, S>) => {
-            this._cleanupSubscription(subscription);
-        });
+        Object
+            .keys(this._handledSubscriptions)
+            .forEach(key => this._cleanupSubscription(this._handledSubscriptions[key]));
```
Comment (ms-markda) [inline src/ComponentBase.ts]: Suggestion: `values(this._handledSubscriptions).forEach(sub => this._cleanupSubscription(sub))`

File: src/ComponentBase.ts [hunk 2]
```diff
@@ -396,18 +378,44 @@ export abstract class ComponentBase<P extends {}, S extends _.Dictionary<any>> e
         return false;
     }
 
+    private _findMatchingSubscription(subscriptions: Dictionary<StoreSubscriptionInternal<P, S>>): boolean {
+        const subscriptionsKeys = Object.keys(subscriptions);
+
+        for (let i = 0, len = subscriptionsKeys.length; i < len; i++) {
```
Comment (ms-markda) [inline src/ComponentBase.ts]: Why use `for` loop here instead of `find` that was being used before?

File: src/tslint/overrideCallsSuperRule.ts [hunk 3]
```diff
@@ -34,8 +33,7 @@ class OverrideCallsSuperWalker extends RuleWalker {
 
     visitMethodDeclaration(node: ts.MethodDeclaration): void {
         const methodName = node.name.getText();
-
-        if (_.includes(this._methodNamesToCheck, methodName)) {
+        if (this._methodNamesToCheck.indexOf(methodName) >= 0) {
```
Comment (ms-markda) [inline src/tslint/overrideCallsSuperRule.ts]: Use `.includes` or shim, similar to `.find`.
```ts
export function includes(array: T[], value: T): boolean {
    if (isFunction(array.includes) {
        return array.includes(value);
    }
    return array.indexOf(value) !== -1;
}
```

File: src/utils.ts [hunk 4]
```diff
@@ -25,6 +45,49 @@ export const assert = (cond: any, message?: string | undefined): void => {
...
-}
+export const values = <T = any>(value: Dictionary<T>): T[] => {
+    if (isFunction(Object.values)) {
+        return Object.values(value);
+    }
+
+    return Object.keys(value).map(key => value[key]);
+};
+
+export const flat = <T = any>(value: T[][]): T[] => (
```
Comment (ms-markda) [inline src/utils.ts]: Nit: The name `flatten` is more standard than `flat`

File: src/ComponentBase.ts [hunk 5]
```diff
@@ -396,18 +378,44 @@ export abstract class ComponentBase<P extends {}, S extends _.Dictionary<any>> e
         return false;
     }
 
+    private _findMatchingSubscription(subscriptions: Dictionary<StoreSubscriptionInternal<P, S>>): boolean {
```
Comment (ms-markda) [inline src/ComponentBase.ts]: This method is only used in one place. It seems a bit premature to extract out of its context. For example, the name is confusing. You say "find" but it returns a `boolean`; should it be "has" instead? I am still not sure what "matching" means. More evidence that this logic is not ready to be its own top-level method out of context.

**Comment by ms-markda** [comment]
I am missing a bit of context; why was this change proposed? I can image a few reasons and I am not against it. But this should probably be documented.

**Comment by berickson1** [comment]
Slowly we've been removing lodash from projects, it's a non-trivial size bump to the package & isn't nearly as much of a perf bump in modern browsers.

File: src/tslint/overrideCallsSuperRule.ts (hunk 3 above)
Comment (a-tarasyuk) [reply src/tslint/overrideCallsSuperRule.ts]: I think we don't need to add a new function for `tslint` rules., I think in the Node.JS environment is ok to use `.includes`.

File: src/utils.ts (hunk 4 above)
Comment (a-tarasyuk) [reply src/utils.ts]: [`flat`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flat) is the new array method, I decided that is acceptable to use the same name.

File: src/ComponentBase.ts (hunk 2 above)
Comment (a-tarasyuk) [reply src/ComponentBase.ts]: `_.find` handles `Arrays` and `Objects`. `.find` works only with arrays. I added a new version with `values` + `find`.

File: src/ComponentBase.ts (hunk 5 above)
Comment (a-tarasyuk) [reply src/ComponentBase.ts]: Removed `_findMatchingSubscription`

File: src/utils.ts [hunk 6]
```diff
@@ -25,6 +45,49 @@ export const assert = (cond: any, message?: string | undefined): void => {
     }
 };
 
-export function noop(): void {
-    // noop
-}
+export const values = <T = any>(value: Dictionary<T>): T[] => {
```
[the remark this answers is not in the document]
Comment (a-tarasyuk) [reply src/utils.ts]: Removed `T = any`

**Pull request merged**